astropy/units/equivalencies.py

```diff
@@ -205,145 +215,34 @@ def spectral_density(wav, factor=None):
-    S_nu = cgs.erg / si.Hz / si.cm**2 / si.s / si.sr
-    nu_S_nu = cgs.erg / si.cm**2 / si.s / si.sr
-    la_S_la = nu_S_nu
-    phot_S_la = astrophys.photon / (si.cm**2 * si.s * si.AA * si.sr)
-    phot_S_nu = astrophys.photon / (si.cm**2 * si.s * si.Hz * si.sr)
-
-    # surface brightness (luminosity equiv)
-    SL_nu = cgs.erg / si.s / si.Hz / si.sr
-    SL_la = cgs.erg / si.s / si.angstrom / si.sr
-    nu_SL_nu = cgs.erg / si.s / si.sr
-    la_SL_la = nu_SL_nu
-    phot_SL_la = astrophys.photon / (si.s * si.AA * si.sr)
-    phot_SL_nu = astrophys.photon / (si.s * si.Hz * si.sr)
-
-    def f_la_to_f_nu(x):
-        return x * (wav.to_value(si.AA, spectral()) ** 2 / c_Aps)
-
-    def f_la_from_f_nu(x):
-        return x / (wav.to_value(si.AA, spectral()) ** 2 / c_Aps)
-
-    def f_nu_to_nu_f_nu(x):
-        return x * wav.to_value(si.Hz, spectral())
-
-    def f_nu_from_nu_f_nu(x):
-        return x / wav.to_value(si.Hz, spectral())
-
-    def f_la_to_la_f_la(x):
-        return x * wav.to_value(si.AA, spectral())
-
-    def f_la_from_la_f_la(x):
-        return x / wav.to_value(si.AA, spectral())
-
-    def phot_f_la_to_f_la(x):
-        return hc * x / wav.to_value(si.AA, spectral())
-
-    def phot_f_la_from_f_la(x):
-        return x * wav.to_value(si.AA, spectral()) / hc
-
-    def phot_f_la_to_f_nu(x):
-        return h_cgs * x * wav.to_value(si.AA, spectral())
-
-    def phot_f_la_from_f_nu(x):
-        return x / (wav.to_value(si.AA, spectral()) * h_cgs)
-
-    def phot_f_la_to_phot_f_nu(x):
-        return x * wav.to_value(si.AA, spectral()) ** 2 / c_Aps
-
-    def phot_f_la_from_phot_f_nu(x):
-        return c_Aps * x / wav.to_value(si.AA, spectral()) ** 2
-
-    phot_f_nu_to_f_nu = phot_f_la_to_f_la
-    phot_f_nu_from_f_nu = phot_f_la_from_f_la
-
-    def phot_f_nu_to_f_la(x):
-        return x * hc * c_Aps / wav.to_value(si.AA, spectral()) ** 3
-
-    def phot_f_nu_from_f_la(x):
-        return x * wav.to_value(si.AA, spectral()) ** 3 / (hc * c_Aps)
-
-    # for luminosity density
-    L_nu_to_nu_L_nu = f_nu_to_nu_f_nu
-    L_nu_from_nu_L_nu = f_nu_from_nu_f_nu
-    L_la_to_la_L_la = f_la_to_la_f_la
-    L_la_from_la_L_la = f_la_from_la_f_la
-
-    phot_L_la_to_L_la = phot_f_la_to_f_la
-    phot_L_la_from_L_la = phot_f_la_from_f_la
-    phot_L_la_to_L_nu = phot_f_la_to_f_nu
-    phot_L_la_from_L_nu = phot_f_la_from_f_nu
-    phot_L_la_to_phot_L_nu = phot_f_la_to_phot_f_nu
-    phot_L_la_from_phot_L_nu = phot_f_la_from_phot_f_nu
-    phot_L_nu_to_L_nu = phot_f_nu_to_f_nu
-    phot_L_nu_from_L_nu = phot_f_nu_from_f_nu
-    phot_L_nu_to_L_la = phot_f_nu_to_f_la
-    phot_L_nu_from_L_la = phot_f_nu_from_f_la
+
+    wav_in_AA = wav.to_value(si.AA, equivalencies=spectral())
+
+    def func_factory(factor):
+        return lambda x: x * factor, lambda x: x / factor
+
+    phot_f_from_f, phot_f_to_f = func_factory(wav_in_AA / (cAAs * h_cgs))
+    flux = [
+        (f_la, f_nu, *func_factory(wav_in_AA**2 / cAAs)),
```

Comment:
Answering https://github.com/astropy/astropy/pull/18100#discussion_r2077919428
```suggestion
        (
            f_la,
            f_nu,
            lambda x: x * wav_in_AA**2 / cAAs,
            lambda x: x * cAAs / wav_in_AA**2,
        ),
```

Reply:
Except that these can be listed, like is done before this PR.

```
(f_la, f_nu, *func_factory(wav_in_AA**2 / cAAs)),
```

becomes

```
func1 = lambda ...
func2 = lambda ...
...
func14 = lambda ...

(f_la, f_nu,  func1, func2),
```

Reply:
[According to PEP 8](https://peps.python.org/pep-0008/#programming-recommendations) it should be written as
```python
def func1(x):
   ...

def func2(x):
   ...
```
and so on. But more importantly, we would have to come up with names for these functions and they will be either hideous or uninformative.

Reply:
do the names not exist already?

Reply:
I played around a bit with lazy evaluation and it seems that if `wav` is a scalar or a small array (I tested with 1000 elements) then eagerly computing the factors makes no difference, so we might as well make the code less verbose. But if `wav` has 1 000 000 elements then I would agree that lazy computation would be better. What is the size of `wav` we want to optimize for?

Reply:
We should absolutely assume that the 99.99% use case is that wav is a scalar. I worry we are having a lot of discussion about at best marginal performance cases.  I think we could improve this along the lines of a class that I suggested above, but would suggest that for now we go with the changes proposed here. It does feel substantially more readable!

Reply:
So then what is the source of the slowdown?

Reply:
Actually I know in Jdaviz that this is often used with a whole slew of arrays, so performance hit for a bit of readability improvement is a 👎 from me.